### PR TITLE
ore: add a clone of bytes::Bytes but backed by lgalloc

### PR DESCRIPTION
**Pull request by danhhz**
To set up moving allocations from our s3 library into lgalloc, we introduce a copy of `bytes::Bytes` backed by lgalloc but with the same API. Crucially, this also implements `bytes::Buf`. We could expand the subset of the `bytes::Bytes` API that this implements, but start with the bare minimum for doing something useful.

If we end up needing to fork our s3 client to push usage of this earlier, we'll likely need to pull this out into a separate crate, but for now start it in mz_ore.

This is intentionally not used anywhere yet. It's pulled into a separate PR to keep code reviewer responsibilities less muddled.

Touches #23332

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by danhhz** [comment]
Rebased this on top of #25088. Will rebase that out once it merges

**Comment by danhhz** [comment]
Rebased on top of 25088, now that it's merged. @antiguru friendly ping on the review :)

File: src/ore/src/lgbytes.rs
Comment (antiguru) [inline src/ore/src/lgbytes.rs:29]: Once it lands, we probably want to switch to `Array` from https://github.com/MaterializeInc/materialize/pull/23944/files#diff-b34b0ca4eb3ec3790fcbc7dc4698ab0bd77a02d7d7fb04a83134aff667f4cc62R115. No need to block this PR, though.

**Review by antiguru — APPROVED** [review]
Looks good, thanks!

File: src/ore/src/lgbytes.rs
Comment (danhhz) [reply src/ore/src/lgbytes.rs:29]: I hadn't seen that yet, what's the advantage?

**Review by danhhz — COMMENTED** [review]
TFTR! Merging on green

**Pull request merged**
